plugins/modules/greppy.py

```diff
@@ -0,0 +1,247 @@
+        type: bool
+    findexit:
+        description: Exit right after finding the first occurrence
+        required: false
+        default: false
+        type: bool
+# Specify this value according to your collection
+# in format of namespace.collection.doc_fragment_name
+extends_documentation_fragment:
+    - community.general.greppy
+
+author:
+    - David Peng (@dpengftw)
+'''
+
+EXAMPLES = '''
+- name: Grep for strings in given list but also exclude certain strings
+  community.general.greppy:
+    path: example.txt
+    search:
+      - first string
+      - second string
+    exclude:
+      - third
+      - fourth
+
+- name: Grep for strings and wait for 120 seconds
+  community.general.greppy:
+    path: example.txt
+    search:
+      - first string
+    timeout: 120
+
+- name: Grep for first occurrence then exit
+  community.general.greppy:
+    path: example.txt
+    search:
+      - first occurrence
+    find_exit: yes
+
+'''
+
+RETURN = '''
+# These are examples of possible return values, and in general should use other names for return values.
+output:
+    description: List of lines that matched given search
+    type: list
+    returned: always
+    sample: [
+            "Ex eorum enim scriptis et institutis cum omnis doctrina liberalis, omnis historia.\n",
+            "Esse enim quam vellet iniquus iustus poterat inpune.\n"
+        ]
+
+matches:
+    description: Number of lines that matched
+    type: int
+    returned: always
+    sample: 2
+
+exclude_matches:
+    description: Number of lines that matched the exclusion strings
+    type: int
+    sample: 1
+'''
+
+import re
+import os
+import time
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_bytes, to_native
+
+def run_checks(module):
+    """Do baseline checks before proceeding."""
+    params = module.params
+    path = params['path']
+
+    b_dest = to_bytes(path, errors='surrogate_or_strict')
+    if os.path.isdir(b_dest):
+        module.fail_json(rc=256, msg='Destination %s is a directory !' % path)
+
+    time_to_wait = 3
+    time_counter = 0
+    while not os.path.exists(path):
+        time.sleep(1)
+        time_counter += 1
+        if time_counter > time_to_wait:
+            module.fail_json(rc=256, msg='Destination %s does not exist !' % path)
+
+def search_line(pattern, string):
+    """Find pattern match in string."""
+    if type(string) is bytes:
+        pc = re.compile(bytes(pattern, "utf-8"))
+    else:
+        pc = re.compile(pattern)
+
+    if re.search(pc, string):
+        return string
+    else:
+        return None
+
+def grep_file(filename, patterns, excludes, ignore_case, timeout, find_exit):
+    """Search a single file."""
+    file = open(filename, 'rb')
+    result = {
+        'output': [],
+        'exclude_matches': 0,
+        'matches': 0
+    }
+
+    timeout_start = time.time()
+
+    while time.time() < timeout_start + timeout:
+        line = file.readline()

```

Comment:
If reading a line blocks, how do you control the timeout?

Reply:
You can always use non-blocking IO and `select`, but that's a lot more low-level. Or use multi-threading. That will definitely make life harder for you though.